// xiaoyuer/gift/db.js

var mysql = require("mysql")
    crypto = require('crypto'),
    fs = require("fs"),
    sprintf = require("sprintf").sprintf;

// 连接数据库
var config_json = fs.readFileSync(__dirname+"/../../../shared/db");

var conn = JSON.parse(config_json);
var db;
handleError();

function handleError(){
    db  = mysql.createConnection(conn);
    db.connect(function(err) {
            if ( !err ) {
                console.log("Connected to MySQL");

            } else if ( err ) {
                console.log(err);
            }
        }

    );
    /*
    *断线则重连
     */
    db.on('error', function (err) {
        console.log('db error', err);
        // 如果是连接断开，自动重新连接
        if (err.code == 'PROTOCOL_CONNECTION_LOST') {
            handleError();
        } else {
            throw err;
        }
    });

}

/*
*用户基本信息
 */
function user_info(){
    /*
    *获取用户信息
     */
    this.get_by_openid = function(openid,callback){
        var sql ="SELECT * FROM user_info WHERE openid = '" + openid+"'";
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
                callback("-1");
            }
            else{
                callabck(result[0]);
            }
        })
    }
    /*
    *新增用户
     */
    this.add_user = function(openid,	nickname){
        var sql = "INSERT INTO user_info VALUES('"+openid+"',1,'"+ nickname +"')";
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
            }
        })
    }
    /*
    *增添抽奖机会
     */
    this.change_chances = function(openid,type){
        var sql = "UPDATE user_info SET left_chance = left_chance"+type+"1";
        db.query(sql,function(err,result){
            if(err){
                console.log(err);

            }
        })
    }
}
/*
*活动集合
* {id:,
* releaseTime,
* active
* }
 */
function activities_set(){
    /*
    *获取所有活动
     */
    this.get_all= function(callback){
        var sql = "SELECT * FROM activity_set";
        function get_it(callback) {
            db.query(sql,function(err,result){
                if(err){
                  console.log(err);
                  callback('-1');
                }
                else{
                    callback(result);
                }
            })
        }
    }
    /*
    *获取还在激活状态的活动
     */
    this.get_active = function(callback){
        var sql = "SELECT * FROM activity_set WHERE is_active = 1";
        function get_it(callback) {
            db.query(sql,function(err,result){
                if(err){
                    console.log(err);
                    callback('-1');
                }
                else{
                    callback(result);
                }
            })
        }
    }
    /*
    *将活动改为不可用
     */
    this.cancel_active = function(id){
       var sql = "UPDATE activity_set SET is_active = 0 WHERE activity_id = " +id;
       function cancel_it(){
           db.query(sql,function(err,result){
                 if(err) {
                    console.log(err);
                     cancel_it();
                 }

           });
       }
    }
    /*
    *新建活动
     */
    this.new_activity = function(){
         var sql = "INSERT INTO activity_set VALUES('','"+ new Date().Format("yy-MM-dd hh:mm:ss")+"',0)"
        function create_new(){
            db.query(sql,function(err,result){
                if(err){
                    console.log(err);
                    create_new();
                }
            })
        }
    }
}
/*
*活动概览
 */
function activity_index(){
    var activi_operate = new activities_set();
    /*
    *获取所有的活动
     */
    this.get_all = function(callback){
        var sql ="SELECT * FROM activity_index";
        db.query(sql,function(err,result){
            if(err)
                console.log(err);
            else{
                callback(result);
            }
        })
    }
    /*
    *获取有效期内的活动
     */
    thi.get_effective = function(callback){
        var sql = "SELECT * FROM activity_index WHERE 	activity_finish > NOW()";
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
            }
            else{
                callback(result);
            }
        })
    }
    /*
    *根据id获取活动的概览
     */
    this.get_by_id = function(id,callback){
        var sql = "SELECT * FROM activity_index WHERE activity_id = "+id;
        dn.query(sql,function(err,result){
            if(err){
                console.log(err);
            }
            else if (result[0]){
                    callback(result[0]);
                if(new Date(result[0].activity_finish) < new Date())
                    activi_operate.cancel_active(id);
            }
            else{
                callback('-1');
            }
        })
    }
    /*
    *设定详情
     */
    this.set_index  = function(obj){
        var sql_temp = "UPDATE activity_index SET event_title = '%s'" +
            ",activity_start='%s',activity_finish = '%s',max_time=%d WHERE activity_id="+
            obj.activity_id;
        var sql = sprintf.sprintf(sql_temp,obj.event_title,obj.activity_start,obj.activity_finish,obj.max_time);
        function set_it(){
            db.query(sql,function(err,result){
                if(err){
                    console.log(err);
                    set_it();
                }
            })
        }
    }
}
/*
*活动详细信息
 */

function activity_info(){
    /*
    *获取所有的活动的详情
     */
    this.get_all = function(callback){
         var sql = "SELECT * FROM activity_info";
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
                callback(-1);
            }
            else{
                callback(result);
            }
        })
    }
    /*
     *获取指定ID的活动的详情
     */
    this.get_id = function(id,callback){
        var sql = "SELECT * FROM activity_info WHERE activity_id = " + id;
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
                callback(-1);
            }
            else{

                callback(result[0]);
            }
        })
    }

    /*
    *新增一个活动的活动详情
     */
    this.new_info = function(obj){
        var sql_temp = "INSERT INTO activity_info VALUES('','%s','%s','%s',%d)";
        var sql = sprintf.sprintf(sql_temp,obj.title,obj.description,obj.picUrl,obj.is_hire);
        function new_in(){
            db.query(sql,function(err,result){
                if(err){
                    console.log(err);
                    new_in();
                }
            })
        }
    }
    /*
    *修改活动详情
     */
    this.set_info = function(obj){
        var sql_temp = "UPDATE  activity_info SET event_title = '%s', 	event_description= '%s', 	" +
            "picUrl = '%s',is_hire = %d WHERE activity_id = "
            +obj.activity_id;
        var sql = sprintf.sprintf(sql_temp,obj.title,obj.description,obj.picUrl,new Date(obj.activity_start).Format("yy-MM-dd hh:mm:ss"),
            new Date(obj.activity_finish).Format("yy-MM-dd hh:mm:ss"),obj.max_time,obj.is_hire);
        function new_in(){
            db.query(sql,function(err,result){
                if(err){
                    console.log(err);
                    new_in();
                }
            })
        }
    }
}

/*
*奖品集合设置
 */
function award_set(){
    /*
    *新增一个抽奖
     */
    this.add_award = function(asso,callback){
        var sql_temp = "INSERT INTO  award_set VALUES('','%s',1,%d)" ;
        var now = new Date().Format("yy-MM-dd hh:mm:ss");
        var sql = sprintf.sprintf(sql_temp,now,asso?asso:"");
        function add_it(){
            db.query(sql,function(err,result){
                if(err){
                    console.log(err);
                    add_it();
                }
                else{
                    db.query("SELECT id FROM award_set WHERE release_time= '" + now+"'",function(err,result){
                        if(err){
                            console.log(err);
                        }
                        else{
                            callback(result[0]);
                        }
                    })
                }
            })

            }
        }
    /*
    *取消一个抽奖的激活
     */
    this.disable = function(id){
        var sql = "UPDATE award_set SET is_active = 0 WHERE id = " + id;
        function dis_it(){
            db.query(sql,function(err,result){
                if(err){
                    console.log(err);
                    dis_it();
                }
            })
        }
    }
    /*
    *获取进行中的抽奖
     */
    this.get_active = function(callback){
        var sql = "SELECT * FROM award_set WHERE is_active = 1 ";
        db.query(sql,function(err,result){
            if(err)
                console.log(err);
            else
                callback(result);
        })
    }
}
/*
*奖品详情
 */
function award_info(){
    /*
    *新增加抽奖活动
    * param: obj[id,{awardlevel,award_name,awrad_left,	award_prob}]
     */
     this.new_award = function(id,obj){
        var sql_temp = "INSERT INTO award_" + id +" VALUES('',%d,'%s',%d,%f,'%s')";
         obj.content.forEach(function(value){
             var sql = sprintf.sprintf(sql_temp,obj.award_level,obj.award_name,obj.award_left,obj.award_prob,obj.award_type);
             db.query(sql,function(err,result){
                 if(err){
                     console.log(err);
                 }
             })
         })
     }
    /*
    *根据id获取奖品信息
     */
    this.get_award = function (id){
        var sql = "SELECT * FROM award_" + id;
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
            }
            else{
                callback(result);
            }
        })
    }
    /*
    *修改指定id中的奖品内容
     */
    this.set_award = function(id,obj){
        var sql_temp = "UPDATE award_" + id + "SET award_level = %d, award_name = '%s',	awrad_left =%d," +
            " award_prob = %f ,award_type= '%s' WHERE the_id=%d";
        var sql   = sprintf.sprintf(sql_temp,obj.award_level,obj.award_name,obj.award_left,obj.award_prob,obj.the_id,obj.award_type);
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
            }
        })
    }
    /*
    *中奖之后的处理
     */
    this.after_luck = function(id,award_level){
        var sql = "UPDATE award_" + id + "SET awrad_left=award_left-1 WHERE award_level = " + award_level;
        function minus_it(){
            db.query(sql,function(err,result){
                 if(err){
                     console.log();
                     minus_it();
                 }
            })
        }
    }
}
/*
*转发流水
 */
function forwarding(){
    /*
    *根据活动id获得该活动转发流水
     */
     this.get_by_id = function(id,callback){
          var sql = "SELECT * FROM forwarding_" + id;
         db.query(sql,function(err,result){
             if(err)
                console.log(err);
             callback(result[0]);
         })
     }
    //新增一个转发
    this.new_one = function(obj){
        var sql_temp = "INSERT INTO forwarding_" + obj.activity_id + "VALUES('%s','%s','%s',1)";
        var sql = sprintf.sprintf(sql_temp,obj.openid,obj.openid_c,new Date().Format("yy-MM-dd hh:mm:ss"));
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
            }
        })
    }
    /*
    *根据openid获取已转发次数
     */
    this.ever_forward = function(openid,id,callback){
        var sql = "SELECT COUNT(*) FROM forwarding_"+ id +"AS total_for WHERE openid  = '" + openid +"'";
        db.query(sql,function(err,result){
            if(err)
                console.log(err);
            callback(result[0].total_for);
        })
    }
}
/*
*抽奖流水
 */
function lottery_result(){
    /*
    *获取所有的抽奖流水
     */
    this.get_all = function(callback){
        var sql = "SELECT * FROM lottery_result";
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
                callback('-1');
            }
            else{
                callback(result);
            }
        })
    }
    /*
    *获取实物奖品抽奖流水
     */
    this.get_real = function(callback){
        var sql = "SELECT * FROM lottery_result WHERE award_type = 'real'";
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
                callback('-1');
            }
            else{
                callback(result);
            }
        })
    }
    /*
    *根据openid获取抽奖结果
     */
    this.get_by_openid = function(openid,callback){
        var sql = "SELECT * FROM lottery_result WHERE openid  ='" +openid+"'" ;
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
            }
            else{
                callback(result);
            }
        })
    }
    /*
    *设定已兑奖
     */
    this.set_finished = function(id){
        var sql= "UPDATE lottery_result SET is_recieved = 1 WHERE lotterty_id = "+id;
        db.query(id,function(err,result){
            if(err){
                db.query(id,function(err,result){
                    if(err){
                       console.log(err);
                    }
                })
                console.log(err);
            }
        })
    }
    /*
    *增加一个抽奖结果
     */
    this.add_result = function(obj){
        var sql_temp = "INSERT INTO lottery_result VALUES('','%s',%d,'%s',%d,'%s','%s')";
        var sql = sprintf.sprintf(sql_temp,obj.openid,obj.is_award,new Date().Format("yy-MM-dd hh:mm:ss"),obj.is_recieved,
            obj.the_code,obj.award_type);
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
            }
        })
    }
}
/*
*实物奖品详情
 */
function lottery_info(){
    this.get_by_id = function(id){
        var sql = "SELECT * FROM lottery_info WHERE the_id =" + id;
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
            }
            else{
                callback(result[0]);
            }

        })
    }
    this.add_one = function(obj){
        var sql_temp = "INSERT INTO lottery_info VALUES('',%d,%d,'%s','%s') ";
        var sql = sprintf.sprintf(sql_temp,obj.award_level,obj.award_id,obj.adress_info,obj.mobile_number);
        db.query(sql,function(err,result){
            if(err)
                console.log(err);
        })
    }
}
/*
*一些砸碎的东西
 */
function util(){
    /*
    *新建转发活动的表
     */
    this.new_forwarding_table = function(id,callback){
         var sql = "CREATE TABLE forwarding_" + id +"LIKE forwarding_temp";
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
                callback('-1');
            }
            else{
                callback('0');
            }
        })
    }
    /*
    *新建奖品的表
     */
    this.new_award_table = function(id,callback){
        var sql =  "CREATE TABLE award_" + id +"LIKE award_temp";
        db.query(sql,function(err,result){
            if(err){
                console.log(err);
                callback('-1');
            }
            else{
                callback('0') ;
            }
        })
    }
}

/*
*获取活动详情
 */
exports.get_event_info = function(event_id,callback){
    var sql;
    sql = "SELECT * FROM event_info WHERE event_id = '" + event_id +"'";
    try{
        db.query(sql,function(err,result){
            if(err)
                throw err;
            callback(result[0]);
        })
    }
    catch(e){
        callback(-1);
        console.log(e);
    }
}
/*
*获取抽奖的流水
 */
exports.get_with_draw_index = function(callback){
    var sql;
    sql = "SELECT * FROM event_withdraw ORDER BY withdraw_generate DESC";
    try{
        db.query(sql,function(err,result){
            if(err)
                throw err;
            callback(result);
        })
    }
    catch(e){
        callback(-1);
        console.log(e);
    }
}

exports.get_with_draw_openid = function(openid,callback){
    var sql;
    sql = "SELECT * FROM event_withdraw WHERE openid = '"+ openid +"' ORDER BY withdraw_generate DESC";
    try{
        db.query(sql,function(err,result){
            if(err)
                throw err;
            callback(result);
        })
    }
    catch(e){
        callback(-1);
        console.log(e);
    }
}
 /*
 *注意！！！！！！！！！！！
 * @TODO
  */
exports.new_withdraw = function(openid,is_award,level,code,callback){
       minus_balance(openid);
    var sql_temp = "INSERT INTO event_withdraw VALUES('%s','%s',%d,'%s'%d,%d,'%s')";
    var sql = sprintf.sprintf(new Date().getTime(),openid,is_award,new Date().Format("yy-MM-dd hh:mm:ss"),0,level,code);
    db.query(sql,function(err,result){
        if(err){
            ;
        }
        callback(result);
    })
}
function minus_balance(openid){
    db.query("UPDATE user_info SET 	user_balance=user_balance+1,witjdraw = withdraw+1 WHERE openid = '" +openid+"')");
}
exports.get_award_left=function(callback){
     db.query("SELECT * FROM award",function(err,result){
         callback(result);
     })
}
exports.get_left_award = function(callback){
    db.query("SELECT * FROM award",function(err,result){
        if(err){
            ;
        }
        callback(result);
    })
}
exports.newAward = function(award_level,	award_name,awrad_left,award_prob,type,callback){
    var now =  new Date().getTime()
    db.query("ALTER  TABLE award RENAME TO award " + now);
    db.query("ALTER  TABLE event_withdraw RENAME TO event_withdraw " + now);
    db.query("CREATE TABLE award " );
    db.query("CREAT TABLE event_withdraw ");
    var sql_temp = "INSERT INTO awrd VALUES(%d,'%s',%d,%f,'%s)";
    var sql= sprintf.sprintf(sql_temp,award_level,	award_name,awrad_left,award_prob,type);
    db.query(sql,function(err,result){
        if(err){
             ;
        }
        callback(result);
    })
}
exports.setAward = function(award_level,	award_name,awrad_left,award_prob,type,callback){
    db.query("DELETE * FROM award WHERE award_level=" +award_level)
    var sql_temp = "INSERT INTO awrd VALUES(%d,'%s',%d,%f,'%s)";
    var sql= sprintf.sprintf(sql_temp,ward_level,	award_name,awrad_left,award_prob,type);
    db.query(sql,function(err,result){
        if(err){
            ;
        }
        callback(result);
    })
}
/*
*获取单个活动有效转发流水
 */
exports.get_forwarding_info= function(table_name,callback){
    var sql;
    sql = "SELECT * FROM " + table_name ;
    try{
        db.query(sql,function(err,result){
            if(err)
                throw err;
            callback(result);
        })
    }
    catch (e){
        callback(-1);
        console.log(e);
    }
}

/*
*获取用户的已转发量
 */
exports.get_ever_get = function(table_name,openid,callback){
    var sql =  "SELECT COUNT(*)  AS total_time FROM " + table_name + " WHERE openid='" + openid +"'";
    try{
        db.query(sql,function(err,result){
            if(err) throw err;
            console.log(result);
            callback(result[0]);
        })
    }
    catch (e){
        callback(-1);
        console.log(e);
    }
}
/*
*获取活动已转发数目，以及其名字
 */
exports.get_event_moreinfo = function(table_name,callback){
    var data = {};
    var sql = "SELECT * FROM event_info WHERE event_id = '" + table_name+"'";
    var sql1 = "SELECT COUNT(*)  AS total_time FROM " + table_name;
    try{
        db.query(sql,function(err,result){
            if(err)
                throw err;
            data.title = result[0].event_title;
            db.query(sql1,function(err,result1){
                if(err)
                    throw err;
                data.total = result1[0].total_time;
                db.query("SELECT event_balance FROM event_index WHERE event_id='"+ table_name +"'",function(err,result2){
                    if(err)
                        throw err;
                    data.balance = result2[0].event_balance;
                    callback(data);
                })
            })
        })
    }
    catch(e){
        callback(-1);
        console.log(e);
    }


}
/*
*往user_info添加内容，如果不存在该用户的话
 */
exports.new_user = function(openid,callback){
    var sql_temp = "INSERT INTO user_info(openid,user_balance,withdraw) SELECT '%s',0,0 FROM dual WHERE NOT EXISTS( SELECT * FROM user_info WHERE openid = '%s')";
    var sql = sprintf(sql_temp,openid,openid);
    try{
        db.query(sql,function(err,result){
            if(err) throw err;
            callback(result);
        })
    }
    catch(e){
        callback(-1);
        console.log(e);
    }
}
/*
*获取用户的信息
 */
exports.user_info = function(openid,callback){
    var sql_temp =  "SELECT * FROM user_info WHERE openid = '%s'";
    var sql = sprintf(sql_temp,openid);
    try{
        db.query(sql,function(err,result){
            if(err) throw err;
            callback(result[0]);
        })
    }
    catch(e){
        callback(-1);
        console.log(e);
    }
}
/*
*增加一个活动
 */
exports.add_event =function(info_obj,callback){
     var sql_temp = "INSERT INTO event_index VALUES('%s',%s,0,'%s','%s',%s,'%s')";
     var id = "event_" + (new Date).getTime();
    try{
        var sql = sprintf(sql_temp,id,info_obj.title,info_obj.start_time,info_obj.finish_time,info_obj.max,(info_obj.email?1:""));
        db.query(sql,function(err,result){
            if(err)throw err;
        });
        var sql_temp1 ="INSERT INTO event_info VALUES('%s','%s','%s','%s')";
        var sql1 = sprintf(sql_temp1,id,info_obj.title,info_obj.description,info_obj.picurl)
        db.query(sql1,function(err,result){
            if(err)throw err;
        })
        var sql_temp2 = "CREATE TABLE %s LIKE event_test";
        var sql2 = sprintf(sql_temp2,id);
        db.query(sql2,function(err,result){
            if(err) throw err;
            callback(result);
        })
    }
    catch(e){
        callback(-1);
        console.log(e);
    }

}

/*
*用户新增加有效转发时的处理
 */
exports.new_award = function(openid,openid1,event_id,callback){
    /*
    *获取活动的奖金余额以及允许的最大转发次数
     */
    if(openid !=openid1){
        var sql_temp4 = "SELECT * FROM %s WHERE openid = '%s' AND openid_c = '%s'";
        var sql4 = sprintf(sql_temp4,event_id,openid,openid1);
        try{
            db.query(sql4,function(err,result4){
                if(err) throw err;
                else if(!result4[0]){
                    var sql_temp = "SELECT event_balance,max_time FROM event_index WHERE event_id = '%s'";
                    var sql = sprintf(sql_temp,event_id);
                    db.query(sql,function(err,result){
                        if(err) throw err;
                        var sql_temp1 = "INSERT INTO %s VALUES('%s','%s','%s',%s)";
                        var sql_temp2 = "UPDATE user_info SET user_balance = user_balance + %d WHERE openid ='%s'";
                        var time = new Date().Format("yyyy-MM-dd hh:mm:ss");

                        /*
                         *获取用户已转发的次数
                         * 若已满则返回-1
                         */
                        db.query("SELECT COUNT(openid) AS total_for FROM "+event_id +" WHERE openid = '" +openid+"'",function(err,res2) {
                            /*
                             *如果用户未达到最大次数
                             */
                            if (res2[0].total_for < result[0].max_time) {


                                    var sql1 = sprintf(sql_temp1, event_id, openid,openid1, time,1);
                                    /*
                                     *加入流水
                                     */
                                    db.query(sql1, function (err, res1) {
                                        if (err) throw err;
                                    })

                                    /*
                                     *用户奖金update
                                     */
                                    var sql2 = sprintf(sql_temp2,1,openid);
                                    db.query(sql2,function(err,res1){
                                        if (err) throw err;
                                    })
                                    callback(1);
                                }

                                                        else{
                                callback(-1);
                            }
                        })


                    })
                }
                else
                    callback(-5);
            })
        }
        catch (e){
            callback(-1);
        }

            }
    else{
        callback(-1);
    }
}

/*
*格式化js日期的自定义方法
 */
Date.prototype.Format = function (fmt) { //author: meizz
    var o = {
        "M+": this.getMonth() + 1, //月份
        "d+": this.getDate(), //日
        "h+": this.getHours(), //小时
        "m+": this.getMinutes(), //分
        "s+": this.getSeconds(), //秒
        "q+": Math.floor((this.getMonth() + 3) / 3), //季度
        "S": this.getMilliseconds() //毫秒
    };
    if (/(y+)/.test(fmt)) fmt = fmt.replace(RegExp.$1, (this.getFullYear() + "").substr(4 - RegExp.$1.length));
    for (var k in o)
        if (new RegExp("(" + k + ")").test(fmt)) fmt = fmt.replace(RegExp.$1, (RegExp.$1.length == 1) ? (o[k]) : (("00" + o[k]).substr(("" + o[k]).length)));
    return fmt;
}

function get_random(){
  return (Math.round((2.5*Math.random() + 1.35)/1.5*100))/100;
}
