// xiaoyuer/gift/db.js

var mysql = require("mysql")
    crypto = require('crypto'),
    fs = require("fs"),
    sprintf = require("sprintf").sprintf;


// 连接数据库
var db;
var config_json = fs.readFileSync(__dirname+"/../../../shared/db");

var conn = JSON.parse(config_json);
var db;

handleError();
function handleError(){
    db  = mysql.createConnection(conn);
    db.connect(function(err) {
            if ( !err ) {
                console.log("Connected to MySQL");

            } else if ( err ) {
                console.log(err);
            }
        }

    );
    /*
    *断线则重连
     */
    db.on('error', function (err) {
        console.log('db error', err);
        // 如果是连接断开，自动重新连接
        if (err.code === 'PROTOCOL_CONNECTION_LOST') {
            handleError();
        } else {
            throw err;
        }
    });

}

/*
*获取活动的列表，索引
 */
exports.get_event_index = function (callback){
    var sql;
    sql = "SELECT * FROM event_index ORDER BY event_start DESC";
    db.query(sql,function(err,result){
        if(err)
            throw err;
        callback(result);
    })
}

/*
*获取活动详情
 */
exports.get_event_info = function(event_id,callback){
    var sql;
    sql = "SELECT * FROM event_info WHERE event_id = '" + event_id +"'";
    db.query(sql,function(err,result){
        if(err)
            throw err;
        callback(result[0]);
    })
}
/*
*获取提现的流水
 */
exports.get_with_draw_index = function(callback){
    var sql;
    sql = "SELECT * FROM event_withdraw ORDER BY withdraw_generate DESC";
    db.query(sql,function(err,result){
        if(err)
            throw err;
        callback(result);
    })
}

/*
*获取单个活动有效转发流水
 */
exports.get_forwarding_info= function(table_name,callback){
    var sql;
    sql = "SELECT * FROM " + table_name ;
    db.query(sql,function(err,result){
        if(err)
            throw err;
        callback(result);
    })
}

/*
*获取用户的已转发量
 */
exports.get_ever_get = function(table_name,openid,callback){
    var sql =  "SELECT COUNT(*)  AS total_time FROM " + table_name + " WHERE openid='" + openid +"'";
    db.query(sql,function(err,result){
        if(err) throw err;
        console.log(result);
        callback(result[0]);
    })
}
/*
*获取活动已转发数目，以及其名字
 */
exports.get_event_moreinfo = function(table_name,callback){
    var data = {};
    var sql = "SELECT * FROM event_info WHERE event_id = '" + table_name+"'";
    var sql1 = "SELECT COUNT(*)  AS total_time FROM " + table_name;
    db.query(sql,function(err,result){
        if(err)
            throw err;
        data.title = result[0].event_title;
        db.query(sql1,function(err,result1){
            if(err)
                throw err;
            data.total = result1[0].total_time;
            db.query("SELECT event_balance FROM event_index WHERE event_id='"+ table_name +"'",function(err,result2){
                if(err)
                    throw err;
                data.balance = result2[0].event_balance;
                callback(data);
            })
        })
    })


}
/*
*往user_info添加内容，如果不存在该用户的话
 */
exports.new_user = function(openid,callback){
    var sql_temp = "INSERT INTO user_info(openid,user_balance,withdraw) SELECT '%s',0.0,0.0 FROM dual WHERE NOT EXISTS( SELECT * FROM user_info WHERE openid = '%s')";
    var sql = sprintf(sql_temp,openid,openid);
    db.query(sql,function(err,result){
        if(err) throw err;
        callback(result);
    })
}
/*
*获取用户的信息
 */
exports.user_info = function(openid,callback){
    var sql_temp =  "SELECT * FROM user_info WHERE openid = '%s'";
    var sql = sprintf(sql_temp,openid);
    db.query(sql,function(err,result){
        if(err) throw err;
        callback(result[0]);
    })
}
/*
*增加一个活动
 */
exports.add_event =function(info_obj,callback){
     var sql_temp = "INSERT INTO event_index VALUES('%s','%s',%s,%s,0,'%s','%s',%s,'%s')";
     var id = "event_" + (new Date).getTime();
     var sql = sprintf(sql_temp,id,info_obj.title,info_obj.total_award,info_obj.total_award,info_obj.start_time,info_obj.finish_time,info_obj.max,(info_obj.email?info_obj.email:""));
     db.query(sql,function(err,result){
         if(err)throw err;
     });
    var sql_temp1 ="INSERT INTO event_info VALUES('%s','%s','%s','%s')";
    var sql1 = sprintf(sql_temp1,id,info_obj.title,info_obj.description,info_obj.picurl)
    db.query(sql1,function(err,result){
        if(err)throw err;
    })
    var sql_temp2 = "CREATE TABLE %s LIKE event_test";
    var sql2 = sprintf(sql_temp2,id);
    db.query(sql2,function(err,result){
        if(err) throw err;
        callback(result);
    })
}

/*
*用户新增加有效转发时的处理
 */
exports.new_award = function(openid,openid1,event_id,callback){
    /*
    *获取活动的奖金余额以及允许的最大转发次数
     */
    if(openid !=openid1){
        var sql_temp4 = "SELECT * FROM %s WHERE openid = '%s' AND openid_c = '%s'";
        var sql4 = sprintf(sql_temp4,event_id,openid,openid1);
        db.query(sql4,function(err,result4){
            if(err) throw err;
            else if(!result4[0]){
                var sql_temp = "SELECT event_balance,max_time FROM event_index WHERE event_id = '%s'";
                var sql = sprintf(sql_temp,event_id);
                db.query(sql,function(err,result){
                    if(err) throw err;
                    var sql_temp1 = "INSERT INTO %s VALUES('%s','%s','%s',%s)";
                    var sql_temp2 = "UPDATE user_info SET user_balance = user_balance + %f WHERE openid ='%s'";
                    var time = new Date().Format("yyyy-MM-dd hh:mm:ss");

                    /*
                     *获取用户已转发的次数
                     * 若已满则返回-1
                     */
                    db.query("SELECT COUNT(openid) AS total_for FROM "+event_id +" WHERE openid = '" +openid+"'",function(err,res2) {
                        /*
                         *如果用户未达到最大次数
                         */
                        if (res2[0].total_for < result[0].max_time) {
                            /*
                             *奖金已发光时,返回-2
                             * 主要用于处理奖金所剩无几时的并发
                             */
                            if(result[0].event_balance<=0){
                                callback(-2);
                            }
                            /*
                             *总奖金余额不足三元时全部发出
                             */
                            else if (result[0].event_balance < 3.0) {

                                var sql1 = sprintf(sql_temp1, event_id, openid,openid1, time, result[0].event_balance);
                                /*
                                 *加入流水
                                 * @param 点击链接的openid
                                 * @param 发送链接的openid
                                 */
                                db.query(sql1, function (err, res1) {
                                    if (err) throw err;
                                })
                                db.query("UPDATE event_index SET event_balance =0 WHERE event_id = '" + event_id +"'", function (err, res1) {
                                    if (err) throw err;
                                })
                                /*
                                 *用户奖金update
                                 */
                                var sql2 = sprintf(sql_temp2,res[0].event_balance,openid);
                                db.query(sql2,function(err,res1){
                                    if (err) throw err;
                                })
                                callback(result[0].event_balance);

                            }
                            /*
                             *随机发放1~2元的奖金
                             */
                            else {
                                var random_money =  get_random();
                                var sql1 = sprintf(sql_temp1, event_id, openid,openid1, time,random_money);
                                /*
                                 *加入流水
                                 */
                                db.query(sql1, function (err, res1) {
                                    if (err) throw err;
                                })
                                db.query("UPDATE event_index SET event_balance = " + (result[0].event_balance - random_money) + " WHERE event_id = '" + event_id + "'", function (err, res1) {
                                    if (err) throw err;
                                })
                                /*
                                 *用户奖金update
                                 */
                                var sql2 = sprintf(sql_temp2,random_money,openid);
                                db.query(sql2,function(err,res1){
                                    if (err) throw err;
                                })
                                callback(random_money);
                            }
                        }
                        else{
                            callback(-1);
                        }
                    })


                })
            }
            else
                callback(-5);
        })

            }
    else{
        callback(-1);
    }
}

/*
*格式化js日期的自定义方法
 */
Date.prototype.Format = function (fmt) { //author: meizz
    var o = {
        "M+": this.getMonth() + 1, //月份
        "d+": this.getDate(), //日
        "h+": this.getHours(), //小时
        "m+": this.getMinutes(), //分
        "s+": this.getSeconds(), //秒
        "q+": Math.floor((this.getMonth() + 3) / 3), //季度
        "S": this.getMilliseconds() //毫秒
    };
    if (/(y+)/.test(fmt)) fmt = fmt.replace(RegExp.$1, (this.getFullYear() + "").substr(4 - RegExp.$1.length));
    for (var k in o)
        if (new RegExp("(" + k + ")").test(fmt)) fmt = fmt.replace(RegExp.$1, (RegExp.$1.length == 1) ? (o[k]) : (("00" + o[k]).substr(("" + o[k]).length)));
    return fmt;
}

function get_random(){
  return 2.22;
}
