// xiaoyuer/hire.js
var nodemailer = require('nodemailer');
var fs = require("fs");
var sprintf = require("sprintf").sprintf;

exports.job_remind = (function(){
    var emailConfig = JSON.parse(fs.readFileSync(__dirname+"/../../shared/appConfig")).email;
    var text_temp = "姓名：  %s\n手机：  %s\n邮箱:  %s\n职位:  %s\n";
    var html_temp = "<b>姓名：&nbsp; %s</b><p/><b>手机：&nbsp; %s</b></p><b>邮箱:&nbsp;  %s</b></p><b>职位: &nbsp; %s</b>"
    var transporter = nodemailer.createTransport({
        service: 'Gmail',
        auth: {
            user: 'chuangkehui@gmail.com',
            pass: 'chuangkehui603'
        }
    });
    var mailOptions = {
        from: 'chuangkehui ✔ <chuangkehui@gmail.com>', // sender address
        to: emailConfig.hr_email, // list of receivers
        subject: '新的应聘者', // Subject line
        text: 'Hello world ✔', // plaintext body
        html: '<b>Hello world ✔</b>' // html body
    };
    return function(name,mobile,addr,job){
        mailOptions.text = sprintf(text_temp,name,mobile,addr,job);
        mailOptions.html = sprintf(html_temp,name,mobile,addr,job);
        transporter.sendMail(mailOptions, function(error, info){
            if(error){
                console.log(error);
            }else{
                console.log('Message sent: ' + info.response);
            }
        });
    }

})();
