// xiaoyuer/order.js
var request = require('request'),
    fs = require("fs"),
    req_json = require("request-json");
var baseUrl = JSON.parse(fs.readFileSync(__dirname+"/../../shared/appConfig")).baseUrl;
/*
*获取订单详情
* @param orderid,type
 */

exports.getOrderInfo = (function (){
    var url = "";
    function get_order_info(urla,openid,id,callback) {
        var url = baseUrl + urla + "?openid=" + openid + "&id="+id;
        try {
            request(url, function (error, response, body) {
                if (!error && response.statusCode == 200) {
                    var resultObj = body;
                    try {
                        resultObj = JSON.parse(body);
                    }
                    catch (e) {
                        console.log(e);
                    }
                    if (error) {
                    }
                    else
                        callback(resultObj);
                }
            })
        }
        catch (e) {
            console.log(e);
            callback(JSON.stringify({code: '-1'}))
        }
    }
    return function(orderid,openid,type,callback){
        switch (type){
            case "service":
                url = "/order/servicedetail.action";
                break;
            case "require":
                url = "/order/requiredetail.action";
                break;
            case "w_service":
                url ="/order/wservicedetail.action";
                break;
            case "w_require":
                url = "/order/wrequiredetail.action";
                break;
            case "game":
                url = "/order/gamedetail.action";
                break;
            default :
                callback(JSON.stringify({code:'-1'}));
        }
        if(url)
            get_order_info(url,openid,orderid,callback)
    }


})();

exports.get_order_list = (function(){
    function get_order(urla,openid,page,type,callback){
        var url = baseUrl + urla + "?openid=" + openid + "&type="+type+"&page=" + page;
        try{
            request(url,function(error,response,body){
                if (!error && response.statusCode==200){
                    var resultObj = body;
                    try{
                        resultObj = JSON.parse(body);
                    }
                    catch (e){
                        console.log(e);
                    }
                    if(error)
                    {}
                    else
                        callback(resultObj);
                }
            })
        }
        catch (e){
            console.log(e);
            callback(JSON.stringify({code:'-1'}))
        }
    }
    /*
     * 发布服务订单
     */
    function  get_service_order_1(openid,page,callback){
      try{
          get_order("/order/service.action",openid,page,1,callback);
      }
      catch (e){
          console.log(e);
          callback(JSON.stringify({code:'-1'}));
      }
    }
    /*
    *预约的服务订单
     */
    function  get_service_order_2(openid,page,callback){
       try{
           get_order( "/order/service.action",openid,page,2,callback);
       }
        catch (e){
            console.log(e);
            callback(JSON.stringify({code:'-1'}));
        }
    }
    /*
     *获取需求订单列表
     * 分页获取
     * @param openid,page
     */

    /*
    *获取发布需求订单
     */
    function  get_require_order1 (openid,page,callback){
        var url = "/order/require.action";
        try{
            get_order(url,openid,page,1,callback);
        }
        catch (e){
            console.log(e);
            callback(JSON.stringify({code:'-1'}));
        }
    }
    /*
     *获取参与的需求订单
     */
    function  get_require_order2 (openid,page,callback){
        var url =  "/order/require.action";
        try{
               get_order(url,openid,page,2,callback);
        }
        catch (e){
            console.log(e);
            callback(JSON.stringify({code:'-1'}))
        }
    }

    /*
     *获取发布的公益服务订单列表
     * 分页获取
     * @param openid,page
     */
    function get_welfare_ser_order1  (openid,page,callback) {
        var url = "/order/wservice.action";
        try{
            get_order(url,openid,page,1,callback);
        }
        catch (e){
            console.log(e);
            callback(JSON.stringify({code:'-1'}))
        }
    }
    /*
     *获取参与的公益服务订单列表
     * 分页获取
     * @param openid,page
     */
    function get_welfare_ser_order2  (openid,page,callback) {
        var url = "/order/wservice.action";
        try{
            get_order(url,openid,page,2,callback);
        }
        catch (e){
            console.log(e);
            callback(JSON.stringify({code:'-1'}))
        }
    }
    /*
     *获取发布的公益需求订单列表
     * 分页获取
     * @param openid,page
     */
    function  get_welfare_req_order1(openid,page,callback) {
        var url = "/order/wrequire.action";
        try{
            get_order(url,openid,page,1,callback);
        }
        catch (e){
            console.log(e);
            callback(JSON.stringify({code:'-1'}))
        }
    }
    /*
     *获取参与的公益需求订单列表
     * 分页获取
     * @param openid,page
     */
    function  get_welfare_req_order2(openid,page,callback) {
        var url = "/order/wrequire.action";
        try{
            get_order(url,openid,page,2,callback);
        }
        catch (e){
            console.log(e);
            callback(JSON.stringify({code:'-1'}))
        }
    }
    /*
     *获取发布的大赛订单列表
     * 分页获取
     * @param openid,page
     */
    function get_order_game1(openid,page,callback){
        var url = "/order/game.action";
        try{
            get_order(url,openid,page,1,callback);
        }
        catch (e){
            console.log(e);
            callback(JSON.stringify({code:'-1'}))
        }
    }
    /*
     *获取发参与的大赛订单列表
     * 分页获取
     * @param openid,page
     */
    function get_order_game2(openid,page,callback){
        var url = "/order/game.action";
        try{
            get_order(url,openid,page,2,callback);
        }
        catch (e){
            console.log(e);
            callback(JSON.stringify({code:'-1'}))
        }
    }
    return {
        require1:get_require_order1,
        require2:get_require_order2,
        service1:get_service_order_1,
        service2:get_service_order_2,
        welfare_ser1:get_welfare_ser_order1,
        welfare_ser2:get_welfare_ser_order2,
        welfare_req1:get_welfare_req_order1,
        welfare_req2:get_welfare_req_order2

    }
})();
