// xiaoyuer/seek.js

var request = require('request'),
    fs = require("fs");
var baseUrl = JSON.parse(fs.readFileSync(__dirname+"/../../shared/appConfig")).baseUrl;
/*
*搜索，返回列表
 */
exports.seek = (function(){
      function seek_it(urla,classify,page,callback){
          try{
              url = baseUrl + urla +"?currentPage="+page+ "&sceneServe=1&cashUint=0&provideBill=-1";
              request(url,function(error,response,body){
                  console.log(body);
                  if(error)
                    console.log(error);
                  var resultObj = body;
                  try{
                      resultObj = JSON.parse(body);
                  }
                  catch (e){
                      console.log(e);
                  }
                  callback(resultObj);
              })
          }
          catch (e){
              callback(JSON.stringify({code:"-1"}));
              console.log(e);
          }
      }
    function serv(classify,page,callback){
        var url = "/search/service.action";
        try{
            seek_it(url,classify,page,callback);
        }
        catch (e){
            callback(JSON.stringify({code:"-1"}));
            console.log(e);
        }
    }
    function req(classify,page,callback){
        var url = "/search/require.action";
        try{
            seek_it(url,classify,page,callback);
        }
        catch (e){
            callback(JSON.stringify({code:"-1"}));
            console.log(e);
        }
    }
    function wel_req(classify,page,callback) {
        var url = "/search/wrequire.action";
        try {
            seek_it(url, classify, page, callback);
        }
        catch (e) {
            callback(JSON.stringify({code: "-1"}));
            console.log(e);
        }
    }
    function wel_ser(classify,page,callback) {
        var url = "/search/wservice.action";
        try {
            seek_it(url, classify, page, callback);
        }
        catch (e) {
            callback(JSON.stringify({code: "-1"}));
            console.log(e);
        }
    }
    function games(classify,page,callback) {
        var url = "/search/game.action";
        try {
            seek_it(url, classify, page, callback);
        }
        catch (e) {
            callback(JSON.stringify({code: "-1"}));
            console.log(e);
        }
    }
    return{
        require:req,
        service:serv,
        wel_require:wel_req,
        wel_service:wel_ser,
        games:games
    }
}) ();
/*
*获取详情
 */
exports.info = (function(){
    function my_trim(str) {
        str.replace(/ /g, "");
        str.replace(/   /g, "")
        return str.replace(/\s/g, "");

    }   function get_info(urla,id,callback){
        var url = baseUrl + urla +"?id=" + id;
        try{
            request(url,function(err,response,body){
                if(err)
                    console.log(err);
                callback(my_trim(body));
            })
        }
        catch (e){
            callback(JSON.stringify({code:"-1"}));
            console.log(e);
        }
    }
   return function(type,id,callback){
       var url = "";
       switch (type){
           case "require":
               url = "/info/require.action";
               break;
           case "service":
               url = "/info/service.action";
               break;
           case "wel_require":
               url = "/info/wrequire.action";
               break;
           case "wel_service":
               url = "/info/wservice.action";
               break;
           case "game":
               url = "/info/game.action";
               break;
           default :
               url = "";
               callback(JSON.stringify({code:"-1"}));
       }
       if(url)
            get_info(url,id,callback);
   }
})();
 /*
 *检查用户是否收藏
 * 收藏则返回1
 * 未收藏为0
 * 未绑定返回-1
  */
exports.is_collect=function (id,openid,type,callback){
    var url = baseUrl + "/need/iscollect?id=" + id + "&openid=" + openid+"&type="+type;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error){
               ;
            }
            else if(resultObj.code==0)
                callback(resultObj.status);
        }
    })
}
/*
*按照ID收藏seek
*异常时返回-1即未绑定
 */
exports.collet = function(id,openid,type,callback){
    var url = baseUrl + "/need/collect?id=" + id + "&openid=" + openid+"&type="+type;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error){
                ;
            }
            else
                callback(resultObj.code);
        }
    })
}
