// xiaoyuer/seek.js

var request = require('request'),
    fs = require("fs"),
    area_class = require("xiaoyuer/area_class"),
    request_json = require("request-json");
var baseUrl = JSON.parse(fs.readFileSync(__dirname+"/../../shared/appConfig")).baseUrl;
var client = request_json.newClient(baseUrl+"/");
/*
*搜索，返回列表
 */
exports.seek = (function(){
      function seek_it(urla,classify,page,orderfield,callback){
              var orderfield = orderfield || "";
              var classifyArr = area_class.classify.get_fullname_by_id(classify);

              var url = baseUrl + urla +"?pno="+page+ "&orderfield="+orderfield+"&sceneServe=-1&serverType=-1&provideBill=-1&classes=" + classifyArr;

              request(url,function(error,response,body){
                  if(error)
                    console.log(error);
                  var resultObj = body;
                  try{
                      resultObj = JSON.parse(body);
                  }
                  catch (e){
                      console.log(e);
                  }
                  callback(resultObj);
              })


      }
    function seek_it_wel(urla,classify,page,orderfield,callback){
        var orderfield = orderfield || "";
        var classifyArr = area_class.welfare_calssify.get_name_by_id(classify);

        var url = baseUrl + urla +"?pno="+page+ "&orderfield="+orderfield+"&sceneServe=-1&serverType=-1&provideBill=-1&classes=" + classifyArr;

        request(url,function(error,response,body){
            if(error)
                console.log(error);
            var resultObj = body;
            try{
                resultObj = JSON.parse(body);
            }
            catch (e){
                console.log(e);
            }
            callback(resultObj);
        })


    }
    function serv(classify,page,orderfield,callback){
        var url = "/search/service.action";
        try{
            seek_it(url,classify,page,orderfield,callback);
        }
        catch (e){
            callback(JSON.stringify({code:"-1"}));
            console.log(e);
        }
    }
    function req(classify,page,orderfield,callback){
        var url = "/search/require.action";
      seek_it(url,classify,page,orderfield ,callback);

    }
    function wel_req(classify,page,orderfield,callback) {
        var url = "/search/wrequire.action";
        try {
            seek_it_wel(url, classify, page,orderfield, callback);
        }
        catch (e) {
            callback(JSON.stringify({code: "-1"}));
            console.log(e);
        }
    }
    function wel_ser(classify,page,orderfield,callback) {
        var url = "/search/wservice.action";
        try {
            seek_it_wel(url, classify, page,orderfield, callback);
        }
        catch (e) {
            callback(JSON.stringify({code: "-1"}));
            console.log(e);
        }
    }
    function games(classify,page,orderfield,callback) {
        var urla = "/search/game.action";
        try{
            var url = baseUrl + urla +"?pno="+page+ "&orderfield=";
            request(url,function(error,response,body){
                console.log(body);
                if(error)
                    console.log(error);
                var resultObj = body;
                try{
                    resultObj = JSON.parse(body);
                }
                catch (e){
                    console.log(e);
                }
                callback(resultObj);
            })
        }
        catch (e){
            callback(JSON.stringify({code:"-1"}));
            console.log(e);
        }
    }
    return{
        require:req,
        service:serv,
        wel_require:wel_req,
        wel_service:wel_ser,
        games:games
    }
}) ();
/*
*获取详情
 */
exports.info = (function(){
    function my_trim(str) {
        str.replace(/ /g, "");
        str.replace(/   /g, "")
        return str.replace(/\s/g, "");

    }   function get_info(urla,id,callback){
        var url = baseUrl + urla +"?id=" + id;

        try{
            request(url,function(err,response,body){
                if(err)
                    console.log(err);
                try{
                    JSON.parse(body);
                }
                catch (e){
                    console.log(e);
                    body = my_trim(body);
                }
                callback(body);
            })
        }
        catch (e){
            callback(JSON.stringify({code:"-1"}));
            console.log(e);
        }
    }
   return function(type,id,callback){
       var url = "";
       switch (type){
           case "require":
               url = "/info/require.action";
               break;
           case "service":
               url = "/info/service.action";
               break;
           case "wrequire":
               url = "/info/wrequire.action";
               break;
           case "wservice":
               url = "/info/wservice.action";
               break;
           case "game":
               url = "/info/game.action";
               break;
           default :
               url = "";
               callback(JSON.stringify({code:"-1"}));
       }
       if(url)
            get_info(url,id,callback);
   }
})();
 /*
 *检查用户是否收藏
 * 收藏则返回1
 * 未收藏为0
 * 未绑定返回-1
  */
exports.is_collect=function (id,openid,type,callback){
    var url = "";
    switch (type){
        case "service":
            url = "collect/checkService.action" ;
            break;
        case "require":
            url = "collect/checkRequire.action";
            break;
        default:
            url = "";

    }
    if(url){
        var data = {openid:openid,id:id};
        client.post(url,data,function(error,response,body){
            callback(body);
        })
    }
    else{
        callback(JSON.stringify({code:-5}));//不支持的接口
    }

}
/*
*按照ID收藏seek
*异常时返回-1即未绑定
 */
exports.collet = function(id,openid,type,callback){
    var obj = {
        id:id,
        openid:openid
    }
    client.post("collect/"+type+".action",obj,function(error,response,body){
        if(error){
            ;
        }
        else
            callback(body);
    })

}
