// xiaoyuer/seek.js

var request = require('request'),
    fs = require("fs");
var baseUrl = JSON.parse(fs.readFileSync(__dirname+"/../../shared/appConfig")).baseUrl;

/*
*按照类型获取服务列表
*
 */
function seekService(type,callback){
    var url = baseUrl + "/service/list?type=" + type;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error){

            }
            else
                callback(resultObj);
        }
    })
}

/*
*根据id获取服务信息
 */
function serviceInfo(id,callback){
    var url = baseUrl + "/service/info?id=" + id;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error){

            }
            else
                callback(resultObj);
        }
    })
}

/*
*根据id和openid收藏服务
 */


function seekGame(type,callback){
    var url = baseUrl + "/game/list?type=" + type;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error){

            }
            else
                callback(resultObj);
        }
    })
}

function gameInfo(id,callback){
    var url = baseUrl + "/game/info?id=" + id;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error){

            }
            else
                callback(resultObj);
        }
    })
}


function seekCommonweal(type,callback){
    var url = baseUrl + "/commonweal/list?type=" + type;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error){

            }
            else
                callback(resultObj);
        }
    })
}

function commonwealInfo(id,callback){
    var url = baseUrl + "/commonweal/info?id=" + id;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error){

            }
            else
                callback(resultObj);
        }
    })
}




function seekNeed(type,callback){
    var url = baseUrl + "/need/list?type=" + type;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error){

            }
            else
                callback(resultObj);
        }
    })
}

function needInfo(id,callback){
    var url = baseUrl + "/need/info?id=" + id;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error){

            }
            else
                callback(resultObj);
        }
    })
}

 /*
 *检查用户是否收藏
 * 收藏则返回1
 * 未收藏为0
 * 未绑定返回-1
  */
exports.is_collect=function (id,openid,type,callback){
    var url = baseUrl + "/need/iscollect?id=" + id + "&openid=" + openid+"&type="+type;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error){
               ;
            }
            else if(resultObj.code==0)
                callback(resultObj.status);
        }
    })
}
/*
*按照ID收藏seek
*异常时返回-1即未绑定
 */
exports.collet = function(id,openid,type,callback){
    var url = baseUrl + "/need/collect?id=" + id + "&openid=" + openid+"&type="+type;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error){
                ;
            }
            else
                callback(resultObj.code);
        }
    })
}

exports.needInfo = needInfo;
exports.seekNeed = seekNeed;
exports.commonwealInfo = commonwealInfo;
exports.seekNCommonweal = seekCommonweal
exports.gameInfo = gameInfo;
exports.seekGame = seekGame;
exports.serviceInfo = serviceInfo;
exports.seekService = seekService;
