// xiaoyuer/userInfo.js
var request = require("request"),
    fs = require("fs"),
    request_js = require("request-json");
var baseUrl = JSON.parse(fs.readFileSync(__dirname+"/../../shared/appConfig")).baseUrl;
var client = request_js.newClient(baseUrl+"/");
/*
*获取用户基本信息
* 返回-1则调到登录页面
*
 */
exports.getUserInfo = function getUserInfo(openid,callback){
    var url = baseUrl + "/user/info.action?openid=" + openid;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(error)
                callback(-1)
            else if(resultObj.code==0)
                callback(resultObj);
            else
                callback(-1);
            /*
            若已绑定返回信息{code:"0"}，否则返回{"code":-1}
             */
        }
    })
}
/*
*用户登录
* 并且绑定
* 异常：-5：账户不存在
*       -6：密码错误
 */
exports.login = function(openid,password,mobile,callback){
    var url =  "reg/login.action";
    var formData = {openid:openid,password:password,mobile:mobile}
    client.post(url,formData,function(error,response,body){
        console.log(body);
        if (!error && response.statusCode==200){
            //返回的似乎就是对象
            var resultObj = body;
            if(error)
            {}
            else
                callback(resultObj.code);
        }
    })
}
/*
*检测用户名是否可用
 */
exports.check_nick = function(nick_name,callback){
    var url = baseUrl +"/user/checkUserName.action?username="+nick_name;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = body;
            try{
                resultObj = JSON.parse(body);
            }
            catch (e){
                console.log(e);
            }
            console.log(resultObj);
            if(error)
            {}
            else
                callback(resultObj.code);
        }
    })

}
/*
*检测手机号是否可用
 */
exports.check_mobile= function(phone_number,callback){
    var url = baseUrl + "/user/checkUserMobile.action?mobile="+phone_number;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(error)
            {}
            else
                callback(resultObj.code);
        }
    })
}
/*
*获取验证码
* 成功时返回cert
 */
exports.get_code = function(phone_number,callback){
    var url = baseUrl + "/user/mobileCode.action?mobile="+phone_number;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(error)
            {}

            else
                callback(resultObj);
        }
    })
}
/*
*注册新用户并且绑定
*成功时返回0
 */
exports.register = function(openid,username,password,mobile,callback){
    var url = "reg/register.action";
    var formData = {openid:openid,username:username,password:password,mobile:mobile};
    console.log(formData);
    client.post(url,formData,function(error,response,body){
        if (!error && response.statusCode==200){
            console.log(body);
            var resultObj = body;
            try{
                resultObj = JSON.parse(body);
            }
            catch (e){
                console.log(e);
            }
            if(error)
            {}
            else
                callback(resultObj);
        }
    });
}
/*
*取消openid与小鱼儿账户的绑定
 */
function cancelBind(openid,callback){
    var url = "reg/unbind.action";
    var data = {openid:openid};
    client.post(url,data,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(error)
            {}
            else
                callback(resultObj);
        }
    })
}

exports.cancelBind = cancelBind;
