// xiaoyuer/userInfo.js
var request = require("request"),
    fs = require("fs");
var baseUrl = JSON.parse(fs.readFileSync(__dirname+"/../../shared/appConfig")).baseUrl;
/*
*获取用户基本信息
* 返回-1则调到登录页面
*
 */
function getUserInfo(openid,callback){
    var url = baseUrl + "/user/info?openid=" + openid;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error)
                callback(-1)
            else if(resultObj.code==0)
                callback(resultObj);
            else
                callback(-1);
            /*
            若已绑定返回信息{code:"0"}，否则返回{"code":-1}
             */
        }
    })
}
/*
*用户登录
* 并且绑定
* 异常：-5：账户不存在
*       -6：密码错误
 */
exports.login = function(openid,password,email,callback){
    var url = baseUrl + "/user/login";
    var formData = {openid:openid,password:password,email:email}
    request({url:url,form:formData},function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error)
            {}
            else
                callback(resultObj.code);
        }
    })
}
/*
*检测用户名是否可用
 */
exports.check_nick = function(nick_name,callback){
    var url = baseUrl +"/user/check/username?username="+nick_name;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error)
            {}
            else
                callback(resultObj.code);
        }
    })

}
/*
*检测手机号是否可用
 */
exports.check_mobile= function(phone_number,callback){
    var url = baseUrl + "/user/check/mobile?mobile="+phone_number;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error)
            {}
            else
                callback(resultObj.code);
        }
    })
}
/*
*获取验证码
* 成功时返回cert
 */
exports.get_code = function(phone_number,callback){
    var url = baseUrl + "/user/check/cert?mobile="+phone_number;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error)
            {}

            else
                callback(resultObj);
        }
    })
}
/*
*注册新用户并且绑定
*成功时返回0
 */
exports.register = function(openid,username,password,mobile){
    var url = baseUrl + "/user/register";
    var formData = {openid:openid,username:username,password:password,mobile:mobile};
    request.post({url:url,form:formData},function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error)
            {}
            else
                callback(resultObj.code);
        }
    });
}
/*
*取消openid与小鱼儿账户的绑定
 */
function cancelBind(openid,callback){
    var url = baseUrl + "/user/cancel?openid=" + openid ;
    request(url,function(error,response,body){
        if (!error && response.statusCode==200){
            var resultObj = JSON.parse(body);
            if(body.error)
            {}
            else
                callback(resultObj);
        }
    })
}
exports.getUserInfo = getUserInfo;
exports.cancelBind = cancelBind;
